In [1]:
import numpy as np
import pickle
import os

seed = 2023

In [2]:
from nlpsig_networks.scripts.ffn_baseline_functions import (
    histories_baseline_hyperparameter_search
)

In [3]:
output_dir = "client_talk_type_output"
if not os.path.isdir(output_dir):
    os.makedirs(output_dir)

## AnnoMI

In [4]:
%run ../load_anno_mi.py

In [5]:
anno_mi.head()

mi_quality  transcript_id                         topic  utterance_id  \
0       high              0  reducing alcohol consumption             0   
1       high              0  reducing alcohol consumption             1   
2       high              0  reducing alcohol consumption             2   
3       high              0  reducing alcohol consumption             3   
4       high              0  reducing alcohol consumption             4   

  interlocutor timestamp                                     utterance_text  \
0    therapist  00:00:13  Thanks for filling it out. We give this form t...   
1       client  00:00:24                                              Sure.   
2    therapist  00:00:25  So, let's see. It looks that you put-- You dri...   
3       client  00:00:34                                            Mm-hmm.   
4    therapist  00:00:34  -and you usually have three to four drinks whe...   

   annotator_id therapist_input_exists therapist_input_subtype  \
0             3                  False                     NaN   
1             3                    NaN                     NaN   
2             3                   True             information   
3             3                    NaN                     NaN   
4             3                   True             information   

  reflection_exists reflection_subtype question_exists question_subtype  \
0             False                NaN            True             open   
1               NaN                NaN             NaN              NaN   
2             False                NaN           False              NaN   
3               NaN                NaN             NaN              NaN   
4             False                NaN           False              NaN   

  main_therapist_behaviour client_talk_type            datetime  
0                 question              NaN 2023-07-10 00:00:13  
1                      NaN          neutral 2023-07-10 00:00:24  
2          therapist_input              NaN 2023-07-10 00:00:25  
3                      NaN          neutral 2023-07-10 00:00:34  
4          therapist_input              NaN 2023-07-10 00:00:34

In [6]:
with open("../anno_mi_sbert.pkl", "rb") as f:
    sbert_embeddings = pickle.load(f)
    
sbert_embeddings.shape

(13551, 384)

# Baseline: FFN using signatures

First, we dimension reduce these and then take signatures. We use the path signature as input to the FFN for classification.

We want to choose a dimension and signature depth such that the number of terms in the signature is _roughly_ 384 so that it is comparable to the number of features that we used for the previous baseline where we computed the mean of the history. Again, we are concatenating the features we obtain with the current utterance embedding.

In [7]:
num_epochs = 100
hidden_dim_sizes = [[32,32],[64,64],[128,128],[256,256],[512,512]]
dropout_rates = [0.5, 0.2, 0.1]
learning_rates = [1e-3, 1e-4, 5e-4]
seeds = [0, 1, 12, 123, 1234]
loss = "focal"
gamma = 2
validation_metric = "f1"

In [8]:
dim_reduce_methods = ["gaussian_random_projection", "umap"]

## Using log signature

In [9]:
log_signature_dimensions_and_sig_depths = [(28, 2), (10, 3), (6, 4)]

In [10]:
import signatory

[signatory.logsignature_channels(channels, depth)
 for (channels, depth) in log_signature_dimensions_and_sig_depths]

[406, 385, 406]

### Using UMAP

In [ ]:
ffn_logsignature_umap_kfold, best_ffn_logsignature_umap_kfold, _, __ = histories_baseline_hyperparameter_search(
    num_epochs=num_epochs,
    df=anno_mi,
    id_column="transcript_id",
    label_column="client_talk_type",
    embeddings=sbert_embeddings,
    y_data=y_data_client,
    output_dim=output_dim_client,
    hidden_dim_sizes=hidden_dim_sizes,
    dropout_rates=dropout_rates,
    learning_rates=learning_rates,
    use_signatures=True,
    seeds=seeds,
    loss=loss,
    gamma=gamma,
    log_signature=True,
    dim_reduce_methods=["umap"],
    dimension_and_sig_depths=log_signature_dimensions_and_sig_depths,
    path_indices=client_index,
    k_fold=True,
    validation_metric=validation_metric,
    results_output=f"{output_dir}/ffn_logsignature_umap_focal_{gamma}_kfold.csv",
    verbose=False
)

  0%|          | 0/3 [00:00<?, ?it/s]

  0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
ffn_logsignature_umap_kfold.groupby(["dimension", "sig_depth", "hidden_dim", "dropout_rate", "learning_rate"]).mean()

In [ ]:
ffn_logsignature_umap_kfold["f1"].mean()

In [ ]:
np.stack(best_ffn_logsignature_umap_kfold["f1_scores"]).mean(axis=0)

### Using random projections

In [ ]:
for i in range(len(log_signature_dimensions_and_sig_depths)):
    print(f"{i}: i || {log_signature_dimensions_and_sig_depths[i]}")
        ffn_logsignature_umap_kfold, best_ffn_logsignature_grp_kfold, _, __ = histories_baseline_hyperparameter_search(
        num_epochs=num_epochs,
        df=anno_mi,
        id_column="transcript_id",
        label_column="client_talk_type",
        embeddings=sbert_embeddings,
        y_data=y_data_client,
        output_dim=output_dim_client,
        hidden_dim_sizes=hidden_dim_sizes,
        dropout_rates=dropout_rates,
        learning_rates=learning_rates,
        use_signatures=True,
        seeds=seeds,
        loss=loss,
        gamma=gamma,
        log_signature=True,
        dim_reduce_methods=["gaussian_random_projection"],
        dimension_and_sig_depths=[log_signature_dimensions_and_sig_depths[i]],
        path_indices=client_index,
        k_fold=True,
        validation_metric=validation_metric,
        results_output=f"{output_dir}/ffn_logsignature_grp_focal_{gamma}_kfold.csv",
        verbose=False
    )

In [ ]:
ffn_logsignature_umap_kfold.groupby(["dimension", "sig_depth", "hidden_dim", "dropout_rate", "learning_rate"]).mean()

In [ ]:
best_ffn_logsignature_grp_kfold["f1"].mean()

In [ ]:
np.stack(best_ffn_logsignature_grp_kfold["f1_scores"]).mean(axis=0)